### PR TITLE
[SPARK-12537] [SQL] Add option to accept quoting of all character backslash quoting mechanism

### PR DESCRIPTION
**Pull request by Cazen**
We can provides the option to choose JSON parser can be enabled to accept quoting of all character or not.

### CODE REVIEW

**Comment by Cazen** [comment]
recreate pull request(10496->10497)

**Comment by srowen** [comment]
In general you do not need to open a new pull request. But before you make a PR, best to finish the questions on the JIRA. I'm not yet clear why this is regularly needed, or when

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -160,6 +160,8 @@ def json(self, path, schema=None):
                 quotes
             * ``allowNumericLeadingZeros`` (default ``false``): allows leading zeros in numbers \
                 (e.g. 00012)
+            * ``allowBackslashEscapingAnyCharacter`` (default ``false``): allows accepting quoting \
```
Comment (rxin) [inline python/pyspark/sql/readwriter.py]: I'd actually change the default value to true...

File: python/pyspark/sql/readwriter.py [hunk 2]
```diff
@@ -160,6 +160,8 @@ def json(self, path, schema=None):
...
+                of all character using backslash quoting mechanism \
```
Comment (rxin) [inline python/pyspark/sql/readwriter.py]: i don't think you need the trailing \ yet

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala [hunk 3]
```diff
@@ -59,6 +62,8 @@ object JSONOptions {
     allowNumericLeadingZeros =
       parameters.get("allowNumericLeadingZeros").map(_.toBoolean).getOrElse(false),
     allowNonNumericNumbers =
-      parameters.get("allowNonNumericNumbers").map(_.toBoolean).getOrElse(true)
+      parameters.get("allowNonNumericNumbers").map(_.toBoolean).getOrElse(true),
```
Comment (Cazen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: Could I ask one more question?
Does "allowNonNumericNumbers"'s default value should be false?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: "true" is more flexible isn't it?

Comment (Cazen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: Yes but I worried confusing when declare class "allowNonNumericNumbers: Boolean = false"

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: What do you mean?

Comment (Cazen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: I'm sorry for lack of english. I know you are very busy
My intention is
- I agree it's good to allowNonNumericNumbers=true
- But There are "allowNonNumericNumbers=false" in comment or declaring
https://github.com/apache/spark/blob/master/python/pyspark/sql/readwriter.py#L161
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L259
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala#L34
- So I wondered if you were trying to set to false

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: I'm suggesting changing the default value to true, and also update all the documentations (as you pointed out) to true.

Comment (Cazen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: I got it. Thank you!

**Comment by Cazen** [comment]
Changed default value to true and remove \ in comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala [hunk 4]
```diff
@@ -59,6 +62,8 @@ object JSONOptions {
...
+    allowBackslashEscapingAnyCharacter =
+      parameters.get("allowBackslashEscapingAnyCharacter").map(_.toBoolean).getOrElse(true)
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: This default should be false, right?

Comment (Cazen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONOptions.scala]: Happy New Year Owen!

At first, I've created PR with false default.

But Xin advised to me that "I'd actually change the default value to true" in previous comment so I've changed.

If you think it doesn't make sense, please call me again. I will change that immediately

Thank you

**Comment by Cazen** [comment]
Could I ask to run test?

**Comment by rxin** [comment]
I've merged this. Thanks!

**Comment by rxin** [comment]
@Cazen you probably want to add this email to your github profile so the commit shows up under your account: Cazen@korea.com

**Comment by Cazen** [comment]
Thank you @rxin I've added my email address to github profile.
I'm very pleased with your help. Have a good day!

**Pull request closed without merging**